src/lib/cache_lib/impl.ml

```diff
@@ -0,0 +1,236 @@
+open Async_kernel
+open Core_kernel
+open Or_error.Let_syntax
+
+module type Inputs_intf = sig
+  val handle_unconsumed_cache_item :
+    logger:Logger.t -> cache_name:string -> unit
+end
+
+module Make (Inputs : Inputs_intf) : Intf.Main.S = struct
+  module rec Cache : sig
+    include Intf.Cache.S with module Cached := Cached
+
+    val remove : 'elt t -> 'elt -> unit Or_error.t
+  end = struct
+    type 'a t = {name: string; set: 'a Hash_set.t; logger: Logger.t}
+
+    let name {name; _} = name
+
+    let create (type elt) ~name ~logger
+        (module Elt : Hash_set.Elt_plain with type t = elt) : elt t =
+      let set =
+        Hash_set.create ~growth_allowed:true ?size:None (module Elt) ()
+      in
+      {name; set; logger}
+
+    let register t x =
+      let%map () = Hash_set.strict_add t.set x in
+      Cached.create ~logger:t.logger t x
+
+    let mem t x = Hash_set.mem t.set x
+
+    let remove t x = Hash_set.strict_remove t.set x
+  end
+  
+  and Cached : sig
+    include Intf.Cached.S
+
+    val create : logger:Logger.t -> 'elt Cache.t -> 'elt -> ('elt, 'elt) t
+  end = struct
+    type (_, _) t =
+      | Base :
+          { data: 'a
+          ; cache: 'a Cache.t
+          ; mutable consumed: bool }
+          -> ('a, 'a) t
+      | Derivative :
+          { original: 'a
+          ; mutant: 'b
+          ; parent: ('c, 'a) t
+          ; mutable consumed: bool }
```

Comment:
can you comment here about the `consumed` semantics. It wasn't obvious to me initially that you'd consume the parent whenever you transform to make a new value.

Reply:
I think my doccoments on the interface make this clearer now.